syntax = "proto3";

package maelstromdb.protos.raft;

import "log.proto";

enum OpCode {
  REQUEST_VOTE = 0;
  APPEND_ENTRIES = 1;
}

message RequestVote {
  message Request {
    uint32 term = 1;
    string candidateId = 2;
    uint32 lastLogIndex = 3;
    uint32 lastLogTerm = 4;
  }

  message Response {
    uint32 term = 1;
    bool voteGranted = 2;
  }
}

message AppendEntries {
  message Request {
    uint32 term = 1;
    string leaderId = 2;
    uint32 prevLogIndex = 3;
    uint32 prevLogTerm = 4;
    repeated log.LogEntry entries = 5;
    uint32 leaderCommit = 6;
  }

  message Response {
    uint32 term = 1;
    bool success = 2;
  }
}

service RaftService {
  rpc RequestVote (RequestVote.Request) returns (RequestVote.Response) {}
  rpc AppendEntries (AppendEntries.Request) returns (AppendEntries.Response) {}
}

